contracts/CreditVoucher.sol

```diff
@@ -0,0 +1,178 @@
+    string public baseTokenURI;
+
+    mapping(uint256 => Voucher) public vouchers;
+
+    // the nonces mapping is given for replay protection
+    mapping(address => uint256) public sigNonces;
+
+    constructor(
+        address _operator,
+        address _usdc,
+        string memory _uri
+    ) ERC721("CreditVoucher", "CV") EIP712("TRVLCreditVoucher", "1") Ownable(_msgSender()) {
+        if (_operator == address(0) || _usdc == address(0)) {
+            revert ZeroAddress();
+        }
+        operator = _operator;
+        USDC = _usdc;
+        baseTokenURI = _uri;
+
+        // Token transfers are disabled by default, except for minting/burning
+        _pause();
+    }
+
+    function _update(address to, uint256 tokenId, address auth) internal override(ERC721Enumerable) returns (address) {
+        if (paused()) {
+            if (_ownerOf(tokenId) != address(0) || to != address(0)) {
+                revert TransferWhilePaused();
+            }
+        }
+        return super._update(to, tokenId, auth);
+    }
+
+    function _baseURI() internal view override returns (string memory) {
+        return baseTokenURI;
+    }
+
+    /**
+     * @notice Set operator address
+     * @dev    Caller must be CONTRACT OWNER
+     * @param _addr The new operator address
+     */
+    function setOperator(address _addr) external onlyOwner {
+        operator = _addr;
+        emit NewOperator(_addr);
+    }
+
+    /**
+     * @notice Set base token URI
+     * @dev    Caller must be CONTRACT OWNER or OPERATOR
+     * @param _uri Th new token base URI
+     */
+    function setBaseTokenURI(string calldata _uri) external {
+        address msgSender = _msgSender();
+        if (msgSender != owner() && msgSender != operator) {
+            revert Unauthorized();
+        }
+        baseTokenURI = _uri;
+        emit NewBaseTokenURI(_uri);
+    }
+
+    /**
+        @notice Set validity duration
+        @dev Caller must be CONTRACT OWNER
+        @param _duration the new validity duration
+    */
+    function setValidityDuration(uint256 _duration) external onlyOwner {
+        validityDuration = _duration;
+        emit NewValidityDuration(_duration);
+    }
+
+    /**
+     * @notice Disable token transfer
+     * @dev Caller must be CONTRACT OWNER
+     */
+    function pause() external onlyOwner {
+        _pause();
+    }
+
+    /**
+     * @notice Enable token transfer
+     * @dev Caller must be CONTRACT OWNER
+     */
+    function unpause() external onlyOwner {
+        _unpause();
+    }
+
+    /**
+     * @notice Lazy mint a credit voucher token
+     * @dev    Caller can be ANYONE
+     * @param _to The new voucher owner
+     * @param _deadline The deadline timestamp by which the call must be mined for the approve to work
+     * @param _signature The signature provided by token owner
+     */
+    function mint(address _to, uint256 _creditValue, uint256 _deadline, bytes calldata _signature) public {
```

Comment:
Only the owner should be able to mint, and only if there are enough USDC in the reserve.

Reply:
For example, if there are 10100 USDC in the reserve and we have already issued several vouchers worth 10000 USDC in total, we shouldn't be able to mint a new voucher worth more than 100 USDC.

Reply:
I've updated the reserve handling as well. Regarding minting, I’m implementing lazy minting to help save on gas fees for us.

Reply:
OK!